File: serpens/src/single_joint/src/generator/order_generator.cpp
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <limits>

#include <ros/ros.h>

#include "single_joint/order_generator.h"

namespace single_joint {
namespace {

OrderGenerator::OrderGenerator(ros::NodeHandle *nh) {      
    nh_ = nh:
    
    nh_->getParam("/order_generator/theta_ld_min", theta_ld_min_);
    nh_->getParam("/order_generator/heta_ld_max", theta_ld_max_);
    nh_->getParam("/order_generator/theta_ld_resolution", theta_ld_resolution_);
    nh_->getParam("/order_generator/theta_l_min", theta_l_min_);
    nh_->getParam("/order_generator/heta_l_max", theta_l_max_);
    nh_->getParam("/order_generator/theta_l_resolution", theta_l_resolution_);
    nh_->getParam("/order_generator/theta_m_min", theta_m_min_);
    nh_->getParam("/order_generator/theta_m_max", theta_m_max_);
    nh_->getParam("/order_generator/theta_m_resolution", theta_m_resolution_);
    nh_->getParam("/order_generator/theta_m_p_min", theta_m_p_min_);
    nh_->getParam("/order_generator/theta_m_p_max", theta_m_p_max_);
    nh_->getParam("/order_generator/theta_m_p_resolution", theta_m_p_resolution_);
 

    position_order_client_ = nh_->serviceClient<PositionOrder>("position_order");
    position_feedback_server_ = nh_->advertiseService("position_feedback", OrderGenerator::FeedbackHandler);

    position_order_ = NULL;
    position_order_count_ = 0;
    position_feedback_ = NULL;
    position_feedback_count_ = 0;

}

void OrderGenerator::Run() {
    // Only the first time the node is initialized.
    if (position_order_ == NULL && position_feedback_ == NULL) {
        position_order_ = OrderGenerator::Generate();
        OrderGenerator::SendOrder();
    }

    ros::spin();
}

void OrderGenerator::SendOrder() {
    if (position_order_client_.call(&position_order_)) {
        position_order_count_ += 1;
    } else {
        ROS_INFO("The position order service client can't reach the server.");
    }
}

void OrderGenerator::FeedbackHandler(PositionFeedback::Request &req, PositionFeedback::Response &res) {
    position_feedback_->request.stamp = req.stamp;
    position_feedback_->request.msg_id = req.msg_id;
    position_feedback_->request.theta_m_update = req.theta_m_update;
    position_feedback_->request.theta_l_update = req.theta_l_update;
    position_feedback_->request.should_reset = req.should_reset;
    
    ROS_INFO("
        PositionFeedback received :\n
        \tstamp : %d,\n
        \tmsg_id : %d,\n
        \ttheta_m_update: %f,\n
        \ttheta_l_update: %f,\n
        \tshoudle_reset: %d,\n
    ", (int)req.stamp, (int)req.msg_id, (float)req.theta_m_update,
    (float)req.theta_l_update, (int)req.should_reset);

    if (req.should_reset) {
        position_feedback_ = NULL;
        position_order_ = OrderGenerator::Generate();
        OrderGenerator::SendOrder();
    } else {
        position_order_ = OrderGenerator::Next();
        OrderGenerator::SendOrder();
    }
}


PositionOrder OrderGenerator::Generate() {
    /*
    Random initialization of the input parameters.
    - 'theta_ld' : will be fixed during the whole episode ('epsilon' is computed using 'theta_ld' and the moving value of 'theta_l').
    - 'theta_l' : will be randomly initialized (a joint is not necessarily straight at the beginning) but will vary through time.
    - 'theta_m' : same as 'theta_l'.
    - 'theta_m_p' ; we randomly decide of which rotationnal speed we want at the beginning. It may vary through time as well.
    */
    float rand_theta_ld = RandomSample(theta_ld_min_, theta_ld_max_, theta_ld_resolution_);
    float rand_theta_l = RandomSample(theta_l_min_, theta_l_max_, theta_l_resolution_);
    float rand_theta_m = RandomSample(theta_m_min_, theta_m_max_, theta_m_resolution_);
    float rand_theta_m_p = RandomSample(theta_m_p_min_, theta_m_p_max_, theta_m_p_resolution_);
    
    // set the values of the differential parameters.
    float theta_l_p = numeric_limits<float>::max();
    float epsilon = numeric_limits<float>::max();
    float epsilon_p = numeric_limits<float>::max();

    // return the Service to send    
    PositionOrder order;
    order.request.stamp = ros::Time::now();
    order.request.msg_id = position_order_count_;
    order.request.theta_ld = rand_theta_ld;
    order.request.theta_l = rand_theta_l;
    order.request.theta_l_p = theta_l_p;
    order.request.theta_m = rand_theta_m;
    order.request.theta_m_p = rand_theta_m_p;
    order.request.epsilon = epsilon;
    order.request.epsilon_p = epsilon_p;

    return order;
}

PositionOrder OrderGenerator::Next() {

        int delta = sent->stamp.nsec - received->stamp.nsec;
        single_joint_episode_generator::JointOrder next;
        _msg_id_counter += 1;
        next.time = ros::Time::now();
        next.msg_id = _msg_id_counter;
        next.theta_ld = sent->theta_ld; // This is the target parameter so it does not change
        next.theta_l = received->theta_l_update;
        next.theta_l_p = (received->theta_l_update - sent->theta_l) / (float)delta;
        next.theta_m = received->theta_m_update;
        next.theta_m_p = (received->theta_m_update - sent->theta_m) / (float)delta;
        next.epsilon = sent->theta_ld - received->theta_l_update;
        next.epsilon_p = ((sent->theta_ld - received->theta_l_update) - sent->epsilon) / (float)delta;
        return next;

}

} // namespace
} // namespace single_joint

float RandomSample(float min_val, float max_val, float resolution) {
    srand(time(NULL)); //initialize the random seed
    int l = (int)((max_val - min_val) / resolution);
    float dist[l];
    for (int i = 0; i < l; i++) {
        dist[i] =  min_val + (float)i*resolution;
    }
    int rand_idx = rand() % l; //generates a random number between 0 and l
    return dist[l];
}

File: serpens/src/single_joint/scripts/generator/order_generator.py
import sys
import random

import rospy

from single_joint.srv import PositionOrder, PositionFeedback

class OrderGenerator:
    
    def __init__(self):
        self.theta_ld_min = rospy.get_param("/generator/theta_ld_min")
        self.theta_ld_max = rospy.get_param("/generator/theta_ld_max")
        self.theta_ld_resolution = rospy.get_param("/generator/theta_ld_resolution")
        self.theta_l_min = rospy.get_param("/generator/theta_l_min")
        self.theta_l_max = rospy.get_param("/generator/theta_l_max")
        self.theta_l_resolution = rospy.get_param("/generator/theta_l_resolution")
        self.theta_m_min = rospy.get_param("/generator/theta_m_min")
        self.theta_m_max = rospy.get_param("/generator/theta_m_max")
        self.theta_m_resolution = rospy.get_param("/generator/theta_m_resolution")
        self.theta_m_p_min = rospy.get_param("/generator/theta_m_p_min")
        self.theta_m_p_max = rospy.get_param("/generator/theta_m_p_max")
        self.theta_m_p_resolution = rospy.get_param("/generator/theta_m_p_resolution")

        rospy.wait_for_service("position_order")
        self.position_order_client = rospy.ServiceProxy("position_order", PositionOrder)
        
        self.position_feedback_server = rospy.Service("position_feedback", PositionFeedback, self._feedback_handler)

        self.position_order = dict()
        self.position_order_count = 0
        self.position_feedback = dict()
        self.position_feedback_count = 0

        rospy.loginfo("Starting the generator node...")


    def run(self):
        if len(self.position_order) == 0 and len(self.position_feedback) == 0:
            self._generate()
            self._send_order()


    def _send_order(self):
        try:
            res = self.position_order_client(**self.position_order)
        except rospy.ServiceException as e:
            rospy.loginfo("Service call failed: %s"%e)
        rospy.loginfo(f"Order sent: {self.position_order}")


    def _feedback_handler(self, req):
        self.position_feedback["stamp"] = req.stamp
        self.position_feedback["msg_id"] = req.msg_id
        self.position_feedback["theta_m_update"] = req.theta_m_update
        self.position_feedback["theta_l_update"] = req.theta_l_update
        self.position_feedback["should_reset"] = req.should_reset

        rospy.loginfo(
            (
                "Packet received:\n" +
                "\tstamp : %i\n" +
                "\tmsg_id : %d\n" +
                "\ttheta_m_update : %f rad\n" +
                "\ttheta_l_update : %f rad\n" +
                "\tshould_reset : %i\n"
            ),
            req.stamp.nsecs,
            req.msg_id,
            req.theta_m_update,
            req.theta_l_update,
            req.should_reset
        )

        if req.should_reset:
            self.position_feedback = dict()
            self._generate()
        else:
            self._next()
        self._send_order()


    def _generate(self):
        """
        Random initialization of the input parameters.
        - 'theta_ld' : will be fixed during the whole episode ('epsilon' is computed using 'theta_ld' and the moving value of 'theta_l').
        - 'theta_l' : will be randomly initialized (a joint is not necessarily straight at the beginning) but will vary through time.
        - 'theta_m' : same as 'theta_l'.
        - 'theta_m_p' ; we randomly decide of which rotationnal speed we want at the beginning. It may vary through time as well.
        """
        rand_theta_ld = self._random_sample(self.theta_ld_min, self.theta_ld_max, self.theta_ld_resolution)
        rand_theta_l = self._random_sample(self.theta_l_min, self.theta_l_max, self.theta_l_resolution)
        rand_theta_m = self._random_sample(self.theta_m_min, self.theta_m_max, self.theta_m_resolution)
        rand_theta_m_p = self._random_sample(self.theta_m_p_min, self.theta_m_p_max, self.theta_m_p_resolution)

        # set the values of the differential parameters with very big values at first (theoretically infinity)
        theta_l_p = 1e20 
        epsilon = 1e20
        epsilon_p = 1e20

        # return the Service to send
        self.position_order["stamp"] = rospy.get_rostime()
        self.position_order["msg_id"] = self.position_order_count
        self.position_order["theta_ld"] = rand_theta_ld
        self.position_order["theta_l"] = rand_theta_l
        self.position_order["theta_l_p"] = theta_l_p
        self.position_order["theta_m"] = rand_theta_m
        self.position_order["theta_m_p"] = rand_theta_m_p
        self.position_order["epsilon"] = epsilon
        self.position_order["epsilon_p"] = epsilon_p

        rospy.loginfo(f"Order generated : {self.position_order}")


    def _next(self):
        next_position = dict()
        delta = self.position_feedback["stamp"] - self.position_order["stamp"]
        next_position["stamp"] = rospy.get_rostime()
        next_position["msg_id"] = self.position_order_count
        next_position["theta_ld"] = self.position_order["theta_ld"]
        next_position["theta_l"] = self.position_feedback["theta_l_update"]
        next_position["theta_l_p"] = (self.position_feedback["theta_l_update"] - self.position_order["theta_l"]) / delta
        next_position["theta_m"] = self.position_feedback["theta_m_update"]
        next_position["theta_m_p"] = (self.position_feedback["theta_m_update"] - self.position_order["theta_m"]) / delta
        next_position["epsilon"] = abs(self.position_feedback["theta_l_update"] - self.position_order["theta_ld"])
        next_position["epsilon_p"] = (
            abs(self.position_feedback["theta_l_update"] - self.position_order["theta_ld"]) -
            self.position_order["epsilon"]
        ) / delta
        
        # update position_order
        self.position_order = next_position


    def _random_sample(self, min_val, max_val, resolution):
        l = int((max_val - min_val) // resolution)
        dist = []
        for i in range(l):
            dist.append(min_val + i * resolution)
        rand_idx = random.randint(0, l)
        return dist[rand_idx]
